version: "3.7"

services:
  prediction_service:
    build:
      context: ./prediction-service
      dockerfile: Dockerfile
    environment:
      - RABBITMQ_URL=amqp://guest:guest@rabbitmq:5672/
      - TORCHSERVE_HOST=http://torchserve:8080
      - REDIS_HOST=redis
      - REDIS_PORT=6379
      - PREDICTION_TIMEOUT=60
    volumes:
      - ./prediction-service/:/app/
      - ./images/:/images
    ports:
      - 8002:8000
    depends_on:
      - rabbitmq
      - redis
    networks:
      - prediction_network

  rabbitmq:
    image: "rabbitmq:3-management"
    hostname: rabbitmq
    ports:
      - "5672:5672"
      - "15672:15672"
    networks:
      - prediction_network

  redis:
    image: "redis:latest"
    ports:
      - "6379:6379"
    networks:
      - prediction_network

  torchserve:
    image: pytorch/torchserve:latest
    ports:
      - "8080:8080"
      - "8081:8081"
    volumes:
      - ./model_store:/home/model-server/model_store
    environment:
      - MODEL_STORE=/home/model-server/model_store
    command:
      - "torchserve"
      - "--start"
      - "--model-store"
      - "/home/model-server/model_store"
      - "--models"
      - "fastrcnn=fastrcnn.mar"
    networks:
      - prediction_network

  worker:
    build:
      context: ./prediction-service
      dockerfile: Dockerfile
    restart: on-failure
    environment:
      - RABBITMQ_URL=amqp://guest:guest@rabbitmq:5672/
      - TORCHSERVE_HOST=http://torchserve:8080
      - REDIS_HOST=redis
      - REDIS_PORT=6379
    volumes:
      - ./prediction-service/:/app/
      - ./images/:/images
    depends_on:
      - rabbitmq
      - redis
      - prediction_service
    networks:
      - prediction_network
    command: python -m worker.worker

networks:
  prediction_network:
